src/ttsim/shared.py

```diff
@@ -13,23 +14,28 @@
 from _gettsim.config import SUPPORTED_GROUPINGS
 
 if TYPE_CHECKING:
-    import datetime
-
     from ttsim.function_types import PolicyFunction
     from ttsim.typing import (
+        DashedISOString,
         GenericCallable,
         NestedDataDict,
         NestedTTSIMObjectDict,
         QualNameTTSIMFunctionDict,
     )
 
 
-_DASHED_ISO_DATE = re.compile(r"\d{4}-\d{2}-\d{2}")
+_DASHED_ISO_DATE_REGEX = re.compile(r"\d{4}-\d{2}-\d{2}")
 
 
-def validate_dashed_iso_date(date: str | datetime.date):
-    if not _DASHED_ISO_DATE.match(date):
-        raise ValueError(f"Date {date} does not match the format YYYY-MM-DD.")
+def to_datetime(date: datetime.date | DashedISOString):
+    if isinstance(date, datetime.date):
+        return date
+    if isinstance(date, str) and _DASHED_ISO_DATE_REGEX.match(date):
```

Comment:
```suggestion
    if isinstance(date, str) and _DASHED_ISO_DATE_REGEX.fullmatch(date):
```

Else, we match only the beginning of `date`, not the entire thing.